package.json

```diff
@@ -123,6 +123,7 @@
     "yaml-front-matter": "^4.0.0"
   },
   "devDependencies": {
+    "@types/leaflet": "^1.9.3",
```

Comment:
Thanks for making me think about this twice. This is one of a couple solutions, which I'll break down the next comment.

Reply:
First, note that this is the error without 18c809b968e3b5c1d6f8596e39506dd4ecfeac6d:

```
node_modules/leaflet-gesture-handling/dist/leaflet-gesture-handling.d.ts:1:20 - error TS7016: Could not find a declaration file for module 'leaflet'. 'node_modules/leaflet/dist/leaflet-src.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/leaflet` if it exists or add a new declaration (.d.ts) file containing `declare module 'leaflet';`

1 import * as L from "leaflet";
                     ~~~~~~~~~
```

So, it's actually coming from `leaflet-gesture-handling` which is [used in src/map.js](https://github.com/nextstrain/auspice/blob/1190f88d9a2d3fd5da2830d0449338514b3c33ff/src/components/map/map.js#L6).

The error appears at the intersection of several circumstances:

1. `allowJs` is enabled so that our TS files can import JS files (see 6f449861ab4e7c5ec8cac9855d4ba6f442ab5dd7).
2. `noImplicitAny` is enabled through `strict` mode, meaning variables in all TS files must be typed (even if explicitly `any`).

    - If we convert map.js → map.ts, we will see the error directly [here](https://github.com/nextstrain/auspice/blob/1190f88d9a2d3fd5da2830d0449338514b3c33ff/src/components/map/map.js#LL5C1-L5C1) instead of indirectly from `leaflet-gesture-handling`)

3. `skipLibCheck` is disabled so all declaration files (for Auspice (currently none) and from dependencies) can be checked properly.
4. `leaflet-gesture-handling` has [a TS declaration file](https://github.com/elmarquis/Leaflet.GestureHandling/blob/a53f3713aec5a0d016066796e542ee37c180b50f/dist/leaflet-gesture-handling.d.ts).
5. The `leaflet` package alone does not provide a TS declaration file. `@types/leaflet` provides it separately.
6. For some reason, if a TS declaration file is available for an import in a JS file, the TypeScript compiler will use it, even though it's a JS file which shouldn't rely on types.

Reply:
Possible solutions and non-solutions based on each point:

1. (non-solution) Disable `allowJs` - this will get rid of the `leaflet-gesture-handling` error but bring forth many other errors due to not having type declarations for all the JS files imported in the converted TS files (due to `noImplicitAny` being enabled).
2. Disable `noImplicitAny` - this will get rid of the error but also hinder TypeScript conversion which relies on surfacing implicit `any` types for proper typing.
3. Enable `skipLibCheck` - this will get rid of the error but also remove the ability to check our own TS declaration files that might come in the future. Ideally there would be an option to skip checking declaration files in `node_modules` alone, but this [isn't currently possible](https://github.com/microsoft/TypeScript/issues/30511).
4. Install @types/leaflet (18c809b968e3b5c1d6f8596e39506dd4ecfeac6d).
5. (non-solution) I don't know of a way to change this behavior.

All are band-aid solutions for an indirect problem. I think (4) is the best long-term option, since we'll need `@types/leaflet` when the map.js → map.ts conversion happens anyways, and it maintains the strictness of TypeScript.

Reply:
> Why is this necessary, and not types for (say) d3?

This is primarily because of points 4-6. For example, let's take this line:

https://github.com/nextstrain/auspice/blob/1190f88d9a2d3fd5da2830d0449338514b3c33ff/src/components/map/map.js#L12

`d3-selection` does not provide TS declaration files, however map.js is a JS file so it's fine to not have declaration files for imported packages.

If `d3-selection` provided a TS declaration file that then imported something like `d3` whose types are in a separate `@types/d3` package, then we would get an error from the `d3-selection` TS declaration file just like the one above.

Or, if we simply rename `map.js` to `map.ts`, we can see the error on that line looking for `d3-selection`'s missing TS declaration file (which is at [`@types/d3-selection`](https://www.npmjs.com/package/@types/d3-selection)).

Reply:
> The error appears at the intersection of several circumstances:

It sure does! Thanks for the great explanation. Step 6 is a bit unexpected, but I get the general impression that mixed (JS+TS) codebases are not what tsc is optimised for. 

Of the solutions I agree that (5) is the way to go. Would you mind updating the commit message to include your explanation from this thread?

Reply:
Updated the commit message in 18c809b968e3b5c1d6f8596e39506dd4ecfeac6d → 3f19625468d00d6c29033589c8a703bce01fcc44.